entitlement/src/main/java/org/killbill/billing/entitlement/dao/DefaultBlockingStateDao.java

```diff
@@ -258,17 +262,21 @@ public Void inTransaction(final EntitySqlDaoWrapperFactory entitySqlDaoWrapperFa
 
                     final BlockingAggregator currentState = getBlockedStatus(sqlDao, entitySqlDaoWrapperFactory.getHandle(), state.getBlockedId(), state.getType(), bundleId, upToDate, context);
                     if (previousState != null && currentState != null) {
-                        recordBusOrFutureNotificationFromTransaction(entitySqlDaoWrapperFactory,
-                                                                     state.getId(),
-                                                                     state.getEffectiveDate(),
-                                                                     state.getBlockedId(),
-                                                                     state.getType(),
-                                                                     state.getStateName(),
-                                                                     state.getService(),
-                                                                     inserted,
-                                                                     previousState,
-                                                                     currentState,
-                                                                     context);
+                        final boolean isBusEvent = state.getEffectiveDate().compareTo(context.getCreatedDate()) <= 0;
+                        if (!isBusEvent || !groupBusEvents || seqId == 0) {
+                            recordBusOrFutureNotificationFromTransaction(entitySqlDaoWrapperFactory,
+                                                                         state.getId(),
```

Comment:
In case we _group_, I'm not sure we want to send any state, do we? For instance, if we were to block a bunch of add-ons, it's not deterministic which `state` we would get here.

In the case of push notifications, I'd rather be explicit (and pass `null` for instance), so that the caller knows not to rely on the metadata.

Now, in case of internal notifications, does that actually work? Does the internal code relies on these ids (e.g. what happens in the example of all add-ons future blocked)?

Reply:
Agreed with the above, and this is something we could (or could not do) as a separate task.

My question was more about the grouping feature: do you know if it works if we turn it on today (e.g. in my example of add-ons future blocked, there is nothing internally looking at these ids)?

Reply:
> My question was more about the grouping feature: do you know if it works if we turn it on today (e.g. in my example of add-ons future blocked, there is nothing internally looking at these ids)?

The limited testing I did shows that it works, and all tests with the feature enabled pass (I will reverify this).

Looking at the code, only Invoice, Beatrix (-> ExtBusEvent) are consumers of these events.
* Invoice is looking at the **effective date** of the event (for the bus, this `now` for all events), and also [extracts the subscriptionId](https://github.com/killbill/killbill/blob/master/invoice/src/main/java/org/killbill/billing/invoice/InvoiceDispatcher.java#L244) but simply to [get access to the account](https://github.com/killbill/killbill/blob/8f22312a0cc091affb4ae345fafde7b6a76a20d8/invoice/src/main/java/org/killbill/billing/invoice/InvoiceDispatcher.java#L272), somewhat of an old scheme as we now have access to `searchKey1` -> `accountRecordId`.
* For beatrix, we will populate the ext bus event with the subscription ID received, which is of course arbitrary considering we would get one event out of many in this mode. I would say, if you configure your system this way, this is what you get (until we get to add new event types).


Also, note that there is no 'future' issues as the code only optimizes for bus events (additional work could be done to also optimize for (grouping) future notifications (but this is a bit more complicated, not that much).

Reply:
Verified locally all tests pass with `mvn  -Dkillbill.features.bus.optimization=true -Dgpg.skip=true  -Pmysql clean test`